torch/_higher_order_ops/out_dtype.py

```diff
@@ -0,0 +1,167 @@
+
+
+class OutDtypeOperator(HigherOrderOperator):
+    """
+    The out_dtype operator takes an existing ATen functional operator, an
+    `out_dtype` argument, and arguments to the original operator, and executes
+    the original operator and returns a Tensor with the `out_dtype` precision.
+    This operator does not mandate a compute precision so it allows the
+    representation to not be opinionated about the exact implementation.
+
+    The general implementation for all operators will be the following:
+        1. Promote inputs dtypes based on default PyTorch dtype promotion rules,
+            using the dtypes of all input Tensors/Scalars and the `out_dtype`
+            arugument.
+        2. Execute the operator
+        3. Cast the output to `out_dtype`
+    """
+
+
+    def __init__(self):
+        super().__init__("out_dtype")
+        # TODO(ydwu4): Subclassing HigherOrderOperator causes __module__ to
+        # become different (torch._higher_order_ops.out_dtype) which will result
+        # in torch.fx to record the op incorrectly in the graph.
+        self.__module__ = "torch.ops.higher_order"
+
+    def __call__(self, op, output_dtype, *args):
+        if not isinstance(op, torch._ops.OpOverload):
+            raise ValueError("out_dtype's first argument must be an OpOverload")
+        if op._schema.is_mutable:
+            raise ValueError("out_dtype's first argument needs to be a functional operator")
+        if not isinstance(op(*args), torch.Tensor):
+            raise ValueError("out_dtype's can only apply to ops that return a single tensor")
+
+        res = super().__call__(op, output_dtype, *args)
+
+        return res
+
+
+out_dtype = OutDtypeOperator()
+out_dtype.fallthrough(DispatchKey.PythonDispatcher)
+out_dtype.fallthrough(DispatchKey.PythonTLSSnapshot)
+out_dtype.fallthrough(DispatchKey.ADInplaceOrView)
+out_dtype.fallthrough(DispatchKey.BackendSelect)
+out_dtype.fallthrough(DispatchKey.AutocastCPU)
+
+
+def trace_out_dtype(proxy_mode, func_overload, op, output_dtype, *args):
+    with disable_proxy_modes_tracing():
+        # This is a simplified implementation of this operator just for tracing.
+        # Actual implementation may also first promote the arguments
+        out = op(*args).to(dtype=output_dtype)
+
+    node_args = (op, output_dtype, *args)
+    proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, node_args)
+    out_proxy = proxy_mode.tracer.create_proxy(
+        "call_function", func_overload, proxy_args, {}, name="out_dtype"
+    )
+    return track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)
+
+
+@out_dtype.py_impl(DispatchKey.CompositeExplicitAutograd)
+def out_dtype_dense(
+    op: torch._ops.OpOverload,
+    output_dtype: torch.dtype,
+    *args
+):
+    flat_inputs = pytree.tree_flatten(args)[0] + [torch.ones(1, dtype=output_dtype)]
+    promote_dtype: torch.dtype = elementwise_dtypes(
```

Comment:
is out_dtype meant to be used with all PyTorch operators? Some PyTorch operators?

The type promotion logic may get into a problematic situation if you have an operator that intentionally takes in an index tensor, like torch.gather. If we pass it a float and a torch.long index tensor, then this logic would attempt to cast the index tensor to torch.float, which would cause the operation to error out.

The implication may be that you actually want a list of acceptable torch._ops.OpOverload that gets checked for.

Reply:
@jerryzh168 Do we want out_dtype to be restricted to a certain set of ops now? Or are you going to followup with a PR to update with the if statement for _int8_int8_int32_mm/bmm/linear mentioned in the chat?

Reply:
@angelayi to unblock please just go with restricting it to the above ops for now

Reply:
Yeah I think we should enable this only ops that we are explicitly using it for.